### PR TITLE
Recommending probabilistic data structures

### PR DESCRIPTION
**Pull request by LiorKogan**

### CODE REVIEW

File: content/ri/memory-optimizations.md [hunk 1]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
 
 The only trade off of converting big hashes to small hashes is that it increase the complexity in your code.
 
-### Switch to bloom filter or hyperloglog
+### Switch from Set to Bloom filter, cuckoo filter, or HyperLogLog
 
-Unique items can be difficult to count. Usually this means storing every unique item then recalling this information somehow. 
+If you are using large [sets](https://redis.io/docs/data-types/sets/) to solve one of the following problems:
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:279]:
```suggestion
Suppose you are using large [sets](https://redis.io/docs/data-types/sets/) to solve one of the following problems:
```

File: content/ri/memory-optimizations.md [hunk 2]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
 
-Redis sets support this with a single command; however, storing every unique item you want want to count may use a prohibitive amount of memory. If this is the case, consider using a HyperLogLog instead. A HyperLogLog is a probabilistic data structure for counting unique items in a set. HyperLogLogs trade off perfect accuracy for less memory usage.
+- Count the number of unique observations in a stream
+- Check if an observation already appeared in the stream
+- Find the percentage or the number of observations in the stream that are smaller or larger than a given value
 
-Bloom filters help when your set contains a high number of elements and you use the set to determine existence or to eliminate duplicates.  
+and you are ready to trade accuracy with speed and memory usage, consider using one of the following probabilistic data structures:
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:285]:
```suggestion
If you are willing to trade accuracy for speed and memory usage, consider using one of the following probabilistic data structures:
```

File: content/ri/memory-optimizations.md [hunk 3]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
 
-Bloom filters aren't natively supported, but you can find several solutions on top of redis. If you are only using the set to count number of unique elements - like unique ip addresses, unique pages visited by a user etc - then switching to hyperloglog saves significant memory.
+- [HyperLogLog](https://redis.io/docs/data-types/hyperloglogs/) can be used for estimating the number of unique observations in a set.
+- [Bloom filter or cuckoo filter](https://redis.io/docs/stack/bloom/) can be used for checking if an observation has already appeared in the stream (false positive matches are possible, but false negatives are not).
+- [t-digest](https://redis.io/docs/stack/bloom/) can be used for estimating the percentage or the number of observations in the stream that are smaller or larger than a given value.
```
Comment (nermiller) [inline content/ri/memory-optimizations.md]:
```suggestion
- [t-digest](https://redis.io/docs/stack/bloom/) to estimate the percentage or the number of observations in the stream that are smaller or larger than a given value.
```

File: content/ri/memory-optimizations.md [hunk 4]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+1. The results achieved from HyperLogLog are not 100% accurate; they have an approximate standard error of 0.81%.
 1. Hyperloglog only tells you the unique count. It cannot tell you the elements in the set.
 
 For example, if you want to maintain how many unique ipaddresses made an API call today. HyperLogLog tells you
 `46966 unique IPs for today`.
 
-But if you want `Show me those 46966 IP Addresses` — it cannot show you. For that, you need to maintain all the IP Addresses in a set
+But if you want `Show me those 46966 IP Addresses` — it cannot show you. For that, you need to maintain all the IP Addresses in a set.
+
+### Switch from Sorted Set to count-min sketch or top-k
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:305]:
```suggestion
### Switch from a sorted set to count-min sketch or top-k
```

File: content/ri/memory-optimizations.md [hunk 5]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+
+If you are using large [sorted sets](https://redis.io/docs/data-types/sorted-sets/) to solve one of the following problems:
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:307]:
```suggestion
Suppose you are using large [sorted sets](https://redis.io/docs/data-types/sorted-sets/) to solve one of the following problems:
```

File: content/ri/memory-optimizations.md [hunk 6]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
 The only trade off of converting big hashes to small hashes is that it increase the complexity in your code.
 
-### Switch to bloom filter or hyperloglog
+### Switch from Set to Bloom filter, cuckoo filter, or HyperLogLog
 
-Unique items can be difficult to count. Usually this means storing every unique item then recalling this information somehow. 
+If you are using large [sets](https://redis.io/docs/data-types/sets/) to solve one of the following problems:
 
-Redis sets support this with a single command; however, storing every unique item you want want to count may use a prohibitive amount of memory. If this is the case, consider using a HyperLogLog instead. A HyperLogLog is a probabilistic data structure for counting unique items in a set. HyperLogLogs trade off perfect accuracy for less memory usage.
+- Count the number of unique observations in a stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:281]:
```suggestion
- Count the number of unique observations in a stream.
```

File: content/ri/memory-optimizations.md [hunk 7]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+- Check if an observation already appeared in the stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:282]:
```suggestion
- Check if an observation already appeared in the stream.
```

File: content/ri/memory-optimizations.md [hunk 8]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+- Find the fraction or the number of observations in the stream that are smaller or larger than a given value
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:283]:
```suggestion
- Find the fraction or the number of observations in the stream that are smaller or larger than a given value.
```

File: content/ri/memory-optimizations.md [hunk 9]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
 `46966 unique IPs for today`.
 
-But if you want `Show me those 46966 IP Addresses` — it cannot show you. For that, you need to maintain all the IP Addresses in a set
+But if you want `Show me those 46966 IP Addresses` — it cannot show you. For that, you need to maintain all the IP Addresses in a set.
+
+### Switch from Sorted Set to count-min sketch or top-k
+
+If you are using large [sorted sets](https://redis.io/docs/data-types/sorted-sets/) to solve one of the following problems:
+
+- Determine the count of a given observation in a stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:309]:
```suggestion
- Determine the count of a given observation in a stream.
```

File: content/ri/memory-optimizations.md [hunk 10]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+- Maintain a list of the most frequent observations in a stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:310]:
```suggestion
- Maintain a list of the most frequent observations in a stream.
```

File: content/ri/memory-optimizations.md [hunk 11]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+
+and you are ready to trade accuracy with speed and memory usage, consider using one of the following probabilistic data structures:
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:312]:
```suggestion
If you are willing to trade accuracy for speed and memory usage, consider using one of the following probabilistic data structures. You can use:
```

File: content/ri/memory-optimizations.md [hunk 12]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+
+- [count-min sketch]() can be used to estimate the count of a given observation in a stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:314]:
```suggestion
- [count-min sketch]() to estimate the count of a given observation in a stream.
```

File: content/ri/memory-optimizations.md [hunk 13]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
...
+- [top-k]() can be used to maintain a list of the _k_ most frequent observations in a stream
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:315]:
```suggestion
- [top-k]() to maintain a list of the _k_ most frequent observations in a stream.
```

File: content/ri/memory-optimizations.md [hunk 14]
```diff
@@ -272,29 +272,49 @@ If you were to change any one the values, you should have a process for moving y
 
 #### Trade offs
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:273]:
```suggestion
#### Tradeoffs
```

File: content/ri/memory-optimizations.md [hunk 15]
```diff
@@ -274,27 +274,47 @@ If you were to change any one the values, you should have a process for moving y
 
 The only trade off of converting big hashes to small hashes is that it increase the complexity in your code.
 
-### Switch to bloom filter or hyperloglog
+### Switch from Set to Bloom filter, cuckoo filter, or HyperLogLog
```
[the remark this answers is not in the document]
Comment (LiorKogan) [reply content/ri/memory-optimizations.md:277]: If referring to Redis Set data structure

File: content/ri/memory-optimizations.md (hunk 4 above)
Comment (LiorKogan) [reply content/ri/memory-optimizations.md:305]: I'm referring to Redis Sorted Set data structure.

File: content/ri/memory-optimizations.md [hunk 16]
```diff
@@ -270,31 +270,51 @@ You should be consistent about the `total no. of expected elements` and the `sha
...
 
-Redis sets support this with a single command; however, storing every unique item you want want to count may use a prohibitive amount of memory. If this is the case, consider using a HyperLogLog instead. A HyperLogLog is a probabilistic data structure for counting unique items in a set. HyperLogLogs trade off perfect accuracy for less memory usage.
+and you are ready to trade accuracy with speed and memory usage, consider using one of the following probabilistic data structures. You can use:
 
-Bloom filters help when your set contains a high number of elements and you use the set to determine existence or to eliminate duplicates.  
+- [HyperLogLog](https://redis.io/docs/data-types/hyperloglogs/) to estimate the number of unique observations in a set.
+- [Bloom filter or cuckoo filter](https://redis.io/docs/stack/bloom/) to check if an observation has already appeared in the stream. False positive matches are possible, but false negatives are not.
+- [t-digest](https://redis.io/docs/stack/bloom/) can be used for estimating the fraction or the number of observations in the stream that are smaller or larger than a given value.
```
Comment (nermiller) [inline content/ri/memory-optimizations.md:289]:
```suggestion
- [t-digest](https://redis.io/docs/stack/bloom/) to estimate the fraction or the number of observations in the stream that are smaller or larger than a given value.
```

**Review by nermiller — APPROVED** [review]
one addl minor edit

**Comment by nermiller** [comment]
@LiorKogan ok to merge? Not sure how it dropped of my radar. Apologies!

**Comment by LiorKogan** [comment]
Yes, we can merge

**Pull request merged**